### PR TITLE
Concurrent strategy execution

### PR DESCRIPTION
**Pull request by avsek477**
Task: https://app.asana.com/0/1201849173362898/1202265970913795

Depends on: 
- https://github.com/bitfinexcom/bfx-hf-models/pull/24
- https://github.com/bitfinexcom/bfx-hf-strategy-exec/pull/20
- https://github.com/bitfinexcom/bfx-hf-strategy/pull/23

Description:
Support multiple strategies concurrently.

UI Changes required:
- in `bfx-hf-ui/scripts/start-api-server.js`, add startegy path in startHFServer function as: **strategyExecutionPath: ${dir}/strategy-executions.json**

**1. Live execution Status**
```
Live execution status endpoint now returns a object with multiple active concurrent strategies.
```

**2. Start Live execution**
```
Client Request: Same as before
Server Response: [
  'strategy.live_execution_started', 
  strategyMapKey, // unique key for the current running strategy
  { ...strategyOpts, startedOn }
]
```

**3. Stop Live execution**
```
Client Request: ['strategy.execute_stop', authToken, strategyMapKey]
Server Response: [
  'strategy.live_execution_stopped', 
  strategyMapKey, // unique key for the current running strategy
  { ...strategyOpts, startedOn }
]
```

**4. Fetch Past Strategies**
```
Client Request: send([
  'get.past_strategies', 
  authToken, 
  queryOpts // optional: { symbol: { eq: 'tMATIC:USD' }, startedOn: { gte: 1650953562507 }}
])
Server Response: ['data.past_strategies', []]
```

**5. New server events responses**
```
['strategy.opened_position_data', strategyMapKey, openpositionDetails] // emitted when a position is opened
['strategy.closed_position_data', strategyMapKey, closedPositionDetails] // emitted when a position is closed
['strategy.rt_execution_results', strategyMapKey, executionResultsObj] // emitted on each price update
['strategy.live_execution_results', strategyMapKey, executionResultsObj] // emitted when the strategy execution is stopped
```

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -28,7 +28,7 @@
     "bfx-hf-models": "git+https://github.com/avsek477/bfx-hf-models.git#strategy-execution",
     "bfx-hf-models-adapter-lowdb": "git+https://github.com/bitfinexcom/bfx-hf-models-adapter-lowdb.git#v1.0.5",
     "bfx-hf-strategy": "git+https://github.com/bitfinexcom/bfx-hf-strategy.git#v1.2.3",
-    "bfx-hf-strategy-exec": "git+https://github.com/avsek477/bfx-hf-strategy-exec.git#stop-processing-msg-on-stop",
+    "bfx-hf-strategy-exec": "git+https://github.com/avsek477/bfx-hf-strategy-exec.git#emit-real-time-execution-results",
```
Comment (avsek477) [inline package.json:31]: change after merge(https://github.com/bitfinexcom/bfx-hf-strategy-exec/pull/20)

File: package.json [hunk 2]
```diff
@@ -27,7 +27,7 @@
     "bfx-hf-indicators": "git+https://github.com/bitfinexcom/bfx-hf-indicators.git#v2.1.1",
     "bfx-hf-models": "git+https://github.com/avsek477/bfx-hf-models.git#strategy-execution",
     "bfx-hf-models-adapter-lowdb": "git+https://github.com/bitfinexcom/bfx-hf-models-adapter-lowdb.git#v1.0.5",
-    "bfx-hf-strategy": "git+https://github.com/bitfinexcom/bfx-hf-strategy.git#v1.2.3",
+    "bfx-hf-strategy": "git+https://github.com/avsek477/bfx-hf-strategy.git#update-position-details",
```
Comment (avsek477) [inline package.json:30]: change after merge(https://github.com/bitfinexcom/bfx-hf-strategy/pull/23)

File: package.json [hunk 3]
```diff
@@ -28,7 +28,7 @@
     "bfx-hf-models": "git+https://github.com/avsek477/bfx-hf-models.git#strategy-execution",
```
Comment (avsek477) [inline package.json:28]: change after merge(https://github.com/bitfinexcom/bfx-hf-strategy/pull/23)

File: lib/ws_servers/api/handlers/strategy/strategy_manager.js [hunk 4]
```diff
@@ -173,116 +230,243 @@ class StrategyManager {
...
+    })
   }
 
   /**
    * @private
-   * @param {boolean} status
    */
-  _setActiveStatus (status) {
-    this.active = status
+  _unsubscribeChannels (channels, strategyMapKey) {
```
Comment (tarcisiozf) [inline lib/ws_servers/api/handlers/strategy/strategy_manager.js:340]: both `_unsubscribe` and `_unsubscribeChannels` have a lot of conditional branches and corner cases, I would recommend increasing the test coverage for that

File: lib/ws_servers/api/handlers/strategy/strategy_manager.js [hunk 5]
```diff
@@ -34,6 +46,15 @@ class StrategyManager {
    * @returns {Promise<void>}
    */
   async start ({ apiKey, apiSecret, authToken }) {
+    if (this.closeConnectionsTimeout) {
+      clearTimeout(this.closeConnectionsTimeout)
+      this.closeConnectionsTimeout = null
+    }
+
+    if (this.ws) {
+      return
```
[the remark this answers is not in the document]
Comment (avsek477) [reply lib/ws_servers/api/handlers/strategy/strategy_manager.js:55]: This is intentional. It is useful to run multiple strategies at once. The strategy manager closes socket connection if there are no strategies running for more than a minute and nullifies the `ws` object. So, if `ws` is present, that means we don't have to establish a connection and hence just returns.

**Review by tarcisiozf — APPROVED** [review]
lgtm

File: package.json (hunk 1 above)
Comment (vigan-abd) [reply package.json:31]:
```suggestion
    "bfx-hf-strategy-exec": "git+https://github.com/bitfinexcom/bfx-hf-strategy-exec.git#v1.2.3",
```

File: package.json (hunk 2 above)
Comment (vigan-abd) [reply package.json:30]:
```suggestion
    "bfx-hf-strategy": "git+https://github.com/bitfinexcom/bfx-hf-strategy.git#v1.2.4",
```

File: package.json (hunk 3 above)
Comment (vigan-abd) [reply package.json:28]:
```suggestion
    "bfx-hf-models": "git+https://github.com/bitfinexcom/bfx-hf-models.git#v2.6.0",
```
